storage/allocator.go

```diff
@@ -13,75 +13,291 @@
+	}
+	return &storeDesc, nil
+}
+
+// capacityGossipUpdate is a gossip callback triggered whenever capacity
+// information is gossiped. It just tracks keys used for capacity
+// gossip.
+func (a *allocator) capacityGossipUpdate(key string, contentsChanged bool) {
+	a.Lock()
+	defer a.Unlock()
+
+	// Clear the cached store lists on new gossip.
+	a.storeLists = nil
+	if a.capacityKeys == nil {
+		a.capacityKeys = map[string]struct{}{}
+	}
+	a.capacityKeys[key] = struct{}{}
+}
 
-	// Randomly pick a node weighted by capacity.
-	var candidates []*proto.StoreDescriptor
-	var capacityTotal float64
+// Allocate returns a suitable store for a new allocation with the
+// required attributes. Nodes already accommodating existing replicas
+// are ruled out as targets.
+func (a *allocator) Allocate(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	stores, err := a.selectRandom(3, required, existing, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Choose the store with the highest percent capacity available.
+	var highStore *proto.StoreDescriptor
 	for _, s := range stores {
-		if _, ok := usedNodes[s.Node.NodeID]; !ok {
-			candidates = append(candidates, s)
-			capacityTotal += s.Capacity.PercentAvail()
+		if highStore == nil || s.Capacity.PercentAvail() > highStore.Capacity.PercentAvail() {
+			highStore = s
+		}
+	}
+	return highStore, nil
+}
+
+// Rebalance returns a suitable store for a rebalance target with
+// required attributes. Rebalance targets are selected at random
+// from amongst stores which fall within an acceptable multiple
+// of standard deviations from either the capacity or range count
+// means.
+func (a *allocator) Rebalance(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	filter := func(s *proto.StoreDescriptor, count, capacity *stat) bool {
+		// Use counts instead of capacities if the cluster has mean capacity
+		// below a threshold level. This is primarily useful for balancing
+		// load evenly in nascent deployments.
+		if capacity.mean > maxAvailCapacityThreshold {
+			return float64(s.Capacity.RangeCount) < count.mean
+		}
+		minAvailCapacity := capacity.mean * (1 + rebalancePctFromMean)
+		if minAvailCapacityThreshold > minAvailCapacity {
+			minAvailCapacity = minAvailCapacityThreshold
+		}
+		return s.Capacity.PercentAvail() > minAvailCapacity
+	}
+	stores, err := a.selectRandom(1, required, existing, filter)
+	if err != nil {
+		return nil, err
+	}
+	return stores[0], nil
+}
+
+// ShouldRebalance returns whether the specified store is overweight
+// according to the cluster mean and should rebalance a range.
+func (a *allocator) ShouldRebalance(s *proto.StoreDescriptor) bool {
+	a.Lock()
+	defer a.Unlock()
+	sl := a.findStores(*s.CombinedAttrs())
+
+	if sl.capacity.mean > maxAvailCapacityThreshold {
```

Comment:
Nearly, but not near enough. I had it originally as a separate method but required passing a flag in which basically resulted in an if statement inside the method, bifurcating the code paths there. This is actually the cleaner approach.